Add a test recipe for testing the speed command

mattcaswell opened the pull request:
Previously there was no test for the speed command. We just do some simple
testing, running the command with various options to confirm that it doesn't
crash or report errors. We use the new -testmode option to ensure that this
happens quickly and doesn't really run full speed tests.

mattcaswell commented:
I'm hoping this will pull our coveralls % up a bit since there was no coverage at all for the apps/speed.c file.

nhorman commented on test/recipes/20-test_speed.t:
```diff
@@ -0,0 +1,119 @@
+#! /usr/bin/env perl
+# Copyright 2017-2023 The OpenSSL Project Authors. All Rights Reserved.
```
nit: copyright date

nhorman reviewed (approved):
just the copyright nit, otherwise looks good

Sashan reviewed (approved):
looks good to me as far as I can tell.

paulidale commented:
With the copyright date fix.

mattcaswell commented:
Fixed the copyright date issue. Also fixed a memory leak in speed that the test revealed. Also changed speed to not compile in support for DSA if DSA support has also been compiled out of the library (in the same way as we do for lots of other algorithms). Finally numerous fixups in the test itself to skip testing of various options where support for that option is not available for some reason.

Please take another look.

mattcaswell commented:
CIs are still running but looking promising...

tom-cosgrove-arm reviewed (approved):
LGTM

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master branch. Thank you.
